### ISSUE TITLE
java.lang.IllegalStateException: Result is already complete: failed

### ISSUE THREAD

**Issue by bolbat** [eclipse-vertx/vert.x#3416]
### Description

We got new exception after updating Vert.x from **3.9.0** to **3.9.1**

Tested with:
- Netty 4.1.49.Final
- Netty: 4.1.50.Final

On the Vert.x **3.9.0** there is **no** this exception (with Netty 4.1.49.Final/4.1.50.Final).

### StackTrace

```
java.lang.IllegalStateException: Result is already complete: failed
	at io.vertx.core.impl.FutureImpl.complete(FutureImpl.java:119)
	at io.vertx.ext.web.client.impl.HttpContext.lambda$handleSendRequest$15(HttpContext.java:530)
	at io.vertx.core.http.impl.HttpClientRequestImpl.connected(HttpClientRequestImpl.java:513)
	at io.vertx.core.http.impl.HttpClientRequestImpl.lambda$null$5(HttpClientRequestImpl.java:466)
	at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:366)
	at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:43)
	at io.vertx.core.impl.ContextImpl.executeFromIO(ContextImpl.java:229)
	at io.vertx.core.impl.ContextImpl.executeFromIO(ContextImpl.java:221)
	at io.vertx.core.http.impl.HttpClientRequestImpl.lambda$connect$7(HttpClientRequestImpl.java:465)
	at io.vertx.core.impl.FutureImpl.dispatch(FutureImpl.java:105)
	at io.vertx.core.impl.FutureImpl.tryComplete(FutureImpl.java:150)
	at io.vertx.core.impl.FutureImpl.complete(FutureImpl.java:111)
	at io.vertx.core.http.impl.Http1xClientConnection.createStream(Http1xClientConnection.java:847)
	at io.vertx.core.http.impl.HttpClientImpl.lambda$getConnectionForRequest$4(HttpClientImpl.java:1041)
	at io.vertx.core.http.impl.ConnectionManager.lambda$getConnection$7(ConnectionManager.java:158)
	at io.vertx.core.http.impl.pool.Pool.connectSucceeded(Pool.java:380)
	at io.vertx.core.http.impl.pool.Pool.access$500(Pool.java:89)
	at io.vertx.core.http.impl.pool.Pool$Holder.lambda$connect$0(Pool.java:127)
	at io.vertx.core.impl.FutureImpl.dispatch(FutureImpl.java:105)
	at io.vertx.core.impl.FutureImpl.tryComplete(FutureImpl.java:150)
	at io.vertx.core.impl.FutureImpl.complete(FutureImpl.java:111)
	at io.vertx.core.http.impl.HttpChannelConnector.lambda$http1xConnected$3(HttpChannelConnector.java:226)
	at io.vertx.core.net.impl.VertxHandler.setConnection(VertxHandler.java:85)
	at io.vertx.core.net.impl.VertxHandler.handlerAdded(VertxHandler.java:102)
	at io.netty.channel.AbstractChannelHandlerContext.callHandlerAdded(AbstractChannelHandlerContext.java:938)
	at io.netty.channel.DefaultChannelPipeline.callHandlerAdded0(DefaultChannelPipeline.java:609)
	at io.netty.channel.DefaultChannelPipeline.addLast(DefaultChannelPipeline.java:223)
	at io.netty.channel.DefaultChannelPipeline.addLast(DefaultChannelPipeline.java:195)
	at io.vertx.core.http.impl.HttpChannelConnector.http1xConnected(HttpChannelConnector.java:232)
	at io.vertx.core.http.impl.HttpChannelConnector.lambda$doConnect$0(HttpChannelConnector.java:162)
	at io.vertx.core.net.impl.ChannelProvider.lambda$connect$1(ChannelProvider.java:78)
	at io.vertx.core.net.impl.ChannelProvider.connected(ChannelProvider.java:160)
	at io.vertx.core.net.impl.ChannelProvider.lambda$handleConnect$2(ChannelProvider.java:143)
	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:577)
	at io.netty.util.concurrent.DefaultPromise.notifyListeners0(DefaultPromise.java:570)
	at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:549)
	at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:490)
	at io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:615)
	at io.netty.util.concurrent.DefaultPromise.setSuccess0(DefaultPromise.java:604)
	at io.netty.util.concurrent.DefaultPromise.trySuccess(DefaultPromise.java:104)
	at io.netty.channel.DefaultChannelPromise.trySuccess(DefaultChannelPromise.java:84)
	at io.netty.channel.epoll.AbstractEpollChannel$AbstractEpollUnsafe.fulfillConnectPromise(AbstractEpollChannel.java:615)
	at io.netty.channel.epoll.AbstractEpollChannel$AbstractEpollUnsafe.finishConnect(AbstractEpollChannel.java:653)
	at io.netty.channel.epoll.AbstractEpollChannel$AbstractEpollUnsafe.epollOutReady(AbstractEpollChannel.java:529)
	at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:465)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:378)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by bolbat** [comment]
I think this can be related somehow to changes:
- 7781c02529c878f424e97dea2007078905c09f46
or
- 9a886b78ca4fe6d9f353f5a70f111afab6d6a71f

**Comment by vietj** [comment]
@bolbat do you have a reproducer ?

**Comment by vietj** [comment]
I think it might be due to https://github.com/vert-x3/vertx-web/commit/b7a9311ad84533b6403dadda1e971ec516ba20d4

**Comment by bolbat** [comment]
> @bolbat do you have a reproducer ?

No, we got this exceptions on production servers where we have high load (30k+ incoming/50k+ outgoing requests per second per instance)

> I think it might be due to [vert-x3/vertx-web@b7a9311](https://github.com/vert-x3/vertx-web/commit/b7a9311ad84533b6403dadda1e971ec516ba20d4)

Seems that this can be cause

**Comment by vietj** [comment]
I will try reproducing it, more info would be welcome of course (and a reproducer a blessing :-) )

**Comment by vietj** [comment]
the issue might be these lines

```java
req.exceptionHandler(continuation::fail);
req.sendHead(v -> continuation.complete());
```

instead continuation.tryFail and continuation.tryComplete should be called . It seems that a request fails and calls continuation fail and the send head will be called and throw this exception but it is not clear how this happen

**Comment by bolbat** [comment]
We already downgraded Vert.x to 3.9.0

I have done some additional checks 
And see that this exception happens mostly when application just started, they make a lot of outgoing connections, and some of them fail (because not all "external" services to wich we make this connections can't handle a lot of incoming connections in one time), and in this case i see this exceptions

But this observation may be false...

**Comment by vietj** [comment]
thanks for sharing we'll investigate this soon

**Comment by vietj** [comment]
this seems to be a race when the HTTP request headers are sent and the connection is closed the the `sendHead` callback is performed but there are chances that writing the request header failed in the process resulting in calling the exception handler that fails the _continuation_ future that is then completed by the `sendHead` callback.

In `HttpClientRequestImpl`

```java
  private void connected(Handler<HttpVersion> headersHandler, HttpClientStream stream) {
    synchronized (this) {
      ....
      stream.writeHead(method, rawMethod, uri, headers, hostHeader(), chunked, pending, ended, priority, continueHandler, handler); // This write likely fails because connection was closed
      ....
    }
    if (headersHandler != null) {
      // This callback assumes headers were sent but does not report any failure
      headersHandler.handle(stream.version());
    }
  }
```

In master this shall not be the case because:

```java
      stream.writeHead(method, uri, headers, authority(), chunked, pending, ended, priority, handler);
```

**Comment by vietj** [comment]
@bolbat are you setting timeout on your HTTP client requests ?

**Comment by bolbat** [comment]
> @bolbat are you setting timeout on your HTTP client requests ?

Yes, 150-1000ms (different on per request basis)

**Comment by vietj** [comment]
ok I think that is how the issue happens.

in practice the issue we are observing is not harmful beside the logging aspect.

we will investigate how to best address it.

**Comment by zenios** [comment]
30k+ incoming/50k+ outgoing requests per second per instance?

Impressive

**Comment by vietj** [comment]
here is a test that reproduce what is happening in with the client:

```java
  @Repeat(times = 1000)
  @Test
  public void testRace() throws Exception {
    NetServer server = vertx.createNetServer();
    server.connectHandler(so -> {
    });
    CountDownLatch latch = new CountDownLatch(1);
    server.listen(8080, "localhost", onSuccess(v -> {
      latch.countDown();
    }));
    awaitLatch(latch);
    HttpClientRequest req = client.get(8080, "localhost", "/");
    req.handler(resp -> {

    });
    AtomicBoolean completed = new AtomicBoolean();
    req.exceptionHandler(err -> {
      if (completed.compareAndSet(false, true)) {
        testComplete();
      }
    });
    req.sendHead(v -> {
      if (completed.compareAndSet(false, true)) {
        testComplete();
      } else {
        System.out.println("Reproduced");
      }
    });
    req.setTimeout(1);
    await();
  }
```

WebClient falls in the `Reproduced` case due to the timer, the `HttpRequest` fails but the HEAD is still written (but not sent as the connection is closed) and the same promise is completed twice resulting in an exception.

We should have use `tryComplete` and `tryFail` instead in `WebClient`

**Comment by vietj** [comment]
Here is the same test with `WebClient`:

```java
  @Repeat(times = 1000)
  @Test
  public void testRace() throws Exception {
    NetServer server = vertx.createNetServer();
    server.connectHandler(so -> {
    });
    CountDownLatch latch = new CountDownLatch(1);
    server.listen(8080, "localhost", onSuccess(v -> {
      latch.countDown();
    }));
    awaitLatch(latch);
    client
      .get(8080, "localhost", "/")
      .timeout(1)
      .send(ar -> {
      testComplete();
    });
    await();
  }
```

**Comment by vietj** [comment]
Using

```java
    req.exceptionHandler(continuation::tryFail);
    req.sendHead(v -> continuation.tryComplete());
```

indeed fixes the issue. I'll open an issue in vertx-web and fix it.

**Comment by vietj** [comment]
Closing this test as invalid , it has been fixed in 3.9 branch
